tools/opt/opt.cpp

```diff
@@ -0,0 +1,160 @@
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+#include <cstring>
+#include <iostream>
+#include <vector>
+
+#include "opt/ir_loader.h"
+#include "opt/libspirv.hpp"
+#include "opt/pass_manager.h"
+
+using namespace spvtools;
+
+void PrintUsage(const char* program) {
+  printf(
+      R"(%s - Optimize a SPIR-V binary file.
+
+USAGE: %s [options] [<input>] -o <output>
+
+The SPIR-V binary is read from <input>. If no file is specified,
+or if <input> is "-", then the binary is read from standard input.
+if <output> is "-", then the optimized output is written to
+standard output.
+
+NOTE: The optimizer is a work in progress.
+
+Options:
+  --strip-debug-info
+               Remove all debug instructions.
+  -h, --help   Print this help.
+  --version    Display optimizer version information.
+)",
+      program, program);
+}
+
+int main(int argc, char** argv) {
+  const char* in_file = nullptr;
+  const char* out_file = nullptr;
+
+  spv_target_env target_env = SPV_ENV_UNIVERSAL_1_1;
+
+  opt::PassManager pass_manager;
+
+  for (int argi = 1; argi < argc; ++argi) {
+    const char* cur_arg = argv[argi];
+    if ('-' == cur_arg[0]) {
+      if (0 == strcmp(cur_arg, "--version")) {
+        printf("%s\n", spvSoftwareVersionDetailsString());
+        return 0;
+      } else if (0 == strcmp(cur_arg, "--help") || 0 == strcmp(cur_arg, "-h")) {
+        PrintUsage(argv[0]);
+        return 0;
+      } else if (0 == strcmp(cur_arg, "-o")) {
+        if (!out_file && argi + 1 < argc) {
+          out_file = argv[++argi];
+        } else {
+          PrintUsage(argv[0]);
+          return 1;
+        }
+      } else if (0 == strcmp(cur_arg, "--strip-debug-info")) {
+        pass_manager.AddPass<opt::StripDebugInfoPass>();
+      } else if (0 == cur_arg[1]) {
+        // Setting a filename of "-" to indicate stdin.
+        if (!in_file) {
+          in_file = cur_arg;
+        } else {
+          fprintf(stderr, "error: More than one input file specified\n");
+          return 1;
+        }
+      } else {
+        PrintUsage(argv[0]);
+        return 1;
+      }
+    } else {
+      if (!in_file) {
+        in_file = cur_arg;
+      } else {
+        fprintf(stderr, "error: More than one input file specified\n");
+        return 1;
+      }
+    }
+  }
+
+  if (out_file == nullptr) {
+    fprintf(stderr, "error: -o required\n");
+    return 1;
+  }
+
+  std::vector<uint32_t> source;
+  const bool use_file = in_file && strcmp("-", in_file);
+  if (FILE* fp = (use_file ? fopen(in_file, "rb") : stdin)) {
+    uint32_t buf[1024];
+    while (size_t len = fread(buf, sizeof(uint32_t),
+                              sizeof(buf) / sizeof(uint32_t), fp)) {
+      source.insert(source.end(), buf, buf + len);
+    }
```

Comment:
Since we are reading 4-bytes at a time, we might want to make sure that we have read the entire file by the end (i.e. there are no stray bytes after the last word)